### ISSUE TITLE
[CHAPTER]: Team Manual for Code Management

### ISSUE THREAD

**Issue by flor14** [the-turing-way/the-turing-way#4027]
### Title

Team Manual for Code Management

### Guide

Guide for collaboration

### What is the purpose of your chapter?  

We have been working on [code management guidelines](https://stringhinilab.github.io/GitHubProceduresLab/) for Dr. Stringhini's lab, which focuses on epidemiology and public health. These guidelines outline the essential tools and software skills new lab members need when they join. For example, in the [onboarding section of the guidelines](https://stringhinilab.github.io/GitHubProceduresLab/onboarding.html), we include instructions for setting up a GitHub team and installing the necessary software. Additionally, there is [a section](https://stringhinilab.github.io/GitHubProceduresLab/lab-design.html) outlining the lab's needs and how they are translated into specific training topics for new members.  

While this first version of the guidelines is tailored to a specific workgroup, we wonder if a chapter on basic code management recommendations could define the minimum training required for new lab members. These suggestions could adapted to be part of a team manual.  

This topic could be part of a _Team Manual_ sub-chapter, focusing on code management and GitHub usage. Depending on the book's structure, it could be integrated into an existing chapter or presented as a new subsection.

### Why are you proposing it?  

Currently, there are sections in *The Turing Way* book that address these topics separately. We believe it is crucial to ensure that new lab members have access to information about code management from the moment they join the lab. The reasons for this are detailed on the introductory page of the [the first page of the guidelines](https://stringhinilab.github.io/GitHubProceduresLab/).  

### Resources

https://stringhinilab.github.io/GitHubProceduresLab/ and the citations mentioned there. I am sure there are more!

### Who can help?

All those interested in the topic! :) It would be interesting to discuss this at a Book Dash.

### Updates

- [ ] Write chapter outline
- [ ] Add material to the chapter
- [ ] Combine materials into a readable chapter
- [ ] Proofread
- [ ] Create PR and link in the comments
- [ ] Request reviews
- [ ] Address reviews
- [ ] Merge to main branch

**Comment by aleesteele** [comment]
@RichardJActon and @JimMadge - this might be of interest for you both!

**Comment by RichardJActon** [comment]
I read the code management guidelines you linked to - I like them a lot some excellent suggestions.

In a generalized version of this I would suggest not presuming github as the place to host you're lab's code repos. But presenting this as among the first choices a group needs to make and one that somewhat locks them into which ever platform they choose for everything outside of core git functionality. Whilst GitHub is the most popular and has some advantages it is not without it's disadvantages. In gitlab or a Forgejo based code forge like [codeberg](https://codeberg.org/) you have more flexibility to change instances or move to a self-hosted instance. Which ever you choose it can be a good idea to setup a mirror to one of the others to serve as a backup in case anything happens to your primary forge.

**Comment by srtee** [comment]
@flor14 this is an amazing initiative! @RichardJActon great comments and additions! Here's some comments from me towards turning this into a nice Book Dash project with deliverables:

### Meta

- [ ] _Who is this for?_ I am assuming this is for either a group leader or informal senior trying to start a team handbook for _their team_ or _team members_, instead of being _for the team members directly_?
- [ ] _What do we call GitHub?_ I agree with @RichardJActon that the wording should be GitHub-agnostic, as other labs may use other "Git hosts" like GitLab, Codeberg, even self-hosted Forgejo ...
  - "Forge" is the FOSS lingo, but newer coders may not grasp the connection with SourceForge.
  - "Collaborative development environment" is what Wikipedia calls Codeberg, or CDE, playing on IDE, but that's a bit wordy.
  - "Git host" is my personal preference, getting at the essential service provided: hosting Git repos online, with additional collaborative capabilities. Maybe "Git cloud"? "Git platform" is what some others (BitBucket) call themselves but it misses the "online"-y-ness for newer readers
- [ ] _How-to first or Why-do first?_
  - My personal preference is "how-to" -- most people who find TTW are already plenty convinced that they should do better reproducible science and want to get stuck right in.
  - "Why-do" is important within team contexts though since motivating people is good for getting them on-board.

### Material
- [ ] _Templatify_ the "GitHub Procedures" steps: create a template that formal / informal group leaders can quickly individualize for their use by replacing Stringhini Lab-specific information with placeholders
  - [ ] Onboarding
  - [ ] Starting A New Project
  - [ ] Regular Project Workflow
  - [ ] _Collaboration Workflow_ (etiquette on how to comment on others' code?? responsibilities of different org members for resolving disputes?)
  - [ ] Offboarding
- [ ] _Maintaining Team Guidelines And Norms_: _how_ should leaders and team members help each other learn these guidelines and norms? What are less helpful ways that norms are often enforced (e.g. exclaiming "RTFM" is _not_ helpful!) and how can that be replaced with more helpful ways? "Ten Reasons" can be really helpful as an introduction here -- if norm enforcement is counter to the reasons for having guidelines in the first place then that's no good!
- [ ] Team Roles: defining roles like Team Maintainer, 🤩Community Manager🤩, Package Maintainer (with responsibility for a particular sub-package), and so on ... as well as community mechanisms for feedback (maybe too heavyweight?)

### Integration
- [ ] Add link-to-resource into "Reproducible Research" > "Resources" overview https://book.the-turing-way.org/reproducible-research/overview/overview-resources
- [ ] Put resources themselves in "Reproducible Research" > "Version Control" https://book.the-turing-way.org/reproducible-research/vcs
- [ ] Turn the "Who Are We and What Do We Need" (https://stringhinilab.github.io/GitHubProceduresLab/lab-design.html) into a Case Study
- [ ] Add link-to-resource into "Guide For Project Design" (for example "Getting Started" checklist)
- [ ] Add link-to-resource into "Guide For Collaboration" (for example "Managing a New Community And Team")

**Comment by JimMadge** [comment]
Slightly worried about the risk of duplicating information. We should check how much of this material is already in the book. Maybe this is more about curation?

**Comment by flor14** [comment]
Hi! Many thanks, @RichardJActon, @JimMadge, and @aleesteele, for taking the time to read the proposal and share your comments. It’s also great to see you reading this, @RichardJActon , and to interact with you again! I always enjoy our exchanges.

Let’s go step by step.

The first thing I want to clarify is that the manual I created is specific to a particular situation and a specific lab. I don’t think the manual itself is a product that can be broadly applied to other contexts, but rather the idea of **creating a manual for each lab where the criteria for software development in that particular environment can be defined**. My contract in this lab was temporary, and the opportunity to explore and guide the use of other technologies was quite limited.

Part of the idea behind the manual, as it is summarized in this recent blog post: [Ten reasons to implement code management practices early in a research group](https://www.software.ac.uk/news/ten-reasons-implement-code-management-practices-early-research-group), is that adopting software management practices and making certain decisions early on can save a lot of time and be highly beneficial. Many labs leave these decisions up to individual students, which can impact the quality of the software produced and affect its reproducibility. Based on my experience working with some labs, a lab manual can help to address that easily.

@RichardJActon:

- I agree that there is very interesting material in lab manuals that include code. I found this material from [Proper](https://prosper.liverpool.ac.uk/manager-of-researchers-resources/the-pi-network/how-lab-handbooks-can-help-shape-research-culture-in-your-team/) with some examples, which I mentioned in the manual. That said, not all labs that have a lab manual include a section on code management. From what I’ve observed, **labs where the leaders have a background in software development are the ones that typically incorporate this type of information in their [manuals](https://www.compgeolab.org/manual/)**. I think what’s valuable here is providing tools for disciplines where this is not an organic part of their workflow.  @aleesteele: Group leaders without formal training in software management would be the target audience in my opinion.

- I love the idea of defining general criteria for choosing the technologies used in a lab. There are so many options! Having clear guidelines would help many lab leaders who lack a software background to make informed decisions. I would introduce categories such as the technology learning curve, whether the technology is paid or free, its level of adoption within the field, and the degree of reproducibility it allows (we can cite your amazing work here @RichardJActon). 
I would also encourage that the manual includes a brief justification for why a certain technology was chosen. This way, the reasoning behind the decision is documented, making it easier to reassess and potentially reconsider alternatives in the future.

@JimMadge :
Yes, there could be some overlap. There is a book chapter about lab manuals, but nothing specifically about software. This could either be integrated into that chapter or developed as a new one. While I’ve used The Turing Way a lot, I haven’t participated enough to have a clear sense of whether this would fit as a subchapter or a full chapter.

Something else I’ve been considering, thanks to a nutritive conversation with Carlos Martinez (@c-martinez), is the vocabulary we use. Most researchers don’t recognize or understand these practices as "software development." That’s why I chose to use the manual *code management*. I think adding some clarifications in this regard would be useful to better guide researchers. What words should or shouldn’t be used in a manual? Maybe an RSE would think of it as software, but a lab leader without a background in the field might need a brief introduction. How should this be approached to maximize the adoption of software development practices?

Finally, I am cc'ing @rmoqpd5t, as he shared his interest in this conversation with me, and @hykelvinlee42, who gave me great feedback on my blog post and who I suspect might enjoy this thread as well.

Apologies for the long message. **I hope this can lead to a short and concrete contribution to the manual.**

**Comment by hykelvinlee42** [comment]
Hi @flor14, I strongly agree some of the vocabulary we use might be quite foreign to those who don't have a background in software development. This was especially crucial when I helped review the draft of a [Software Management Plan for the Digital Research Alliance of Canada](https://zenodo.org/records/13242504). The document is tailored to all the different stakeholders from different backgrounds. An SMP might be helpful in this regard.  

[https://alliancecan.ca/en/services/research-software/funding-and-initiatives](https://alliancecan.ca/en/services/research-software/funding-and-initiatives)

**Comment by c-martinez** [comment]
Hi @flor14 and all, thanks for including me in this thread, I find it a very interesting topic!

I think there is some overlap with another issue we were working on, [Section on Software Management Plans](https://github.com/the-turing-way/the-turing-way/issues/2419).

Whether it should be a section or a chapter, and whether to call it manual for code management or software management plan, the ideas are very similar: it is important to provide some guidance on how to write code (or scripts or software) in the best way possible (and "the best way possible" depends on the context).

Maybe a good idea to merge both efforts, and try to tackle the issues of vocabulary and scope together.

**Comment by RichardJActon** [comment]
@flor14 I very much like the angle of "a guide for writing a lab handbook for your group's specific needs". This has a scope that extends beyond code management and into domain specific stuff. The example I linked above has a whole section on field work with things like the training you need to have done to work on research boats. We don't need to have anything specific to say about domain specific things but stuff like code management is not the only thing where TTW might have some generically applicable resources, conduct, ethics, accessibility etc. as well as the general idea that your lab manual could provide/point to resources like what training you need to do for specific tasks in the group - be that boats, radioisotope handling, or HPC clusters. Starting out with a "how to write the code management section of your lab handbook" might lead to some these other things being added later on.

Depending on the scale at which your group operates a whole very official sounding "software management plan" can easily come off as overkill and 'paperwork' that might get in the way of people in your group to get things done quickly - even if it would actually be a sensible thing for someone to adopt given what they are doing. I think it is important to try and present people with approachable incremental options. 

When writing a lab handbook you will want to ask questions like this:
Does your group produce and maintain software libraries or pipelines that are used and depended on by others? - Then you should really be thinking about software management plans.
Do people in your group write some one-off data analysis / visualization scripts for specific projects - here's a best practices checklist and a standard project layout template.

**Comment by c-martinez** [comment]
Just a quick reaction to @RichardJActon comment:

> When writing a lab handbook you will want to ask questions like this:
> Does your group produce and maintain software libraries or pipelines that are used and depended on by others? - Then you should really be thinking about software management plans.
> Do people in your group write some one-off data analysis / visualization scripts for specific projects - here's a best practices checklist and a standard project layout template.

I agree! Although in the [Practical guide to Software Management Plans](https://zenodo.org/records/7589725), we talk about the two two examples of software (software libraries or one-off data script) as two different types of software. I would say they BOTH can use an SMP, although with different levels of requirements (a _best practices checklist_ for the script, a more extensive plan for the library).

I would like to change the perception of an SMP being "overkill" and "paperwork", because what we try to say in the guidelines is that it can be very simple, if that fits the needs of your software.

**Comment by srtee** [comment]
I don't think that "data management plan" or "software management plan" is so scary a terminology **if** communicated in a safe, friendly environment, as a way to use minimal effort to ensure some amount of code and data reproducibility.

But if it goes in the "team manual" section (which sounds great!), it could be called a "team data manual" or a "team software manual" instead?

**Comment by RichardJActon** [comment]
@c-martinez & @srtee I'm totally in agreement about DMPs and SMPs being of a size and complexity suitable to the scope of the project, and that for smaller projects the existing guides on writing them do actually do this quite well. 

But I think the presence of the advice on writing such as smaller scope plan in a document that covers both small and large scope projects makes them seem more intimidating than they are in practice. A guide like the one proposed here is a good vector to get people familiar with these concepts as applied to projects at the scale they are working at. If your lab is just producing a few analysis scripts you probably won't seek out a document on SMPs on your own, but you might want to write a lab manual. So if you read a guide for writing lab manuals and encounter a reference to a section in a document on SMPs for projects of the type that you do, then you may be more likely to read the relevant section of the document on SMPs. (Note the reference to a section of a document on SMPs not the whole thing, this at least gives the impression you won't have to read the whole thing.)

So I think this is a good vector to tackle any perceptions that DMPs/SMPs are not applicable to a group's work.

**Comment by c-martinez** [comment]
Ok, so I think we could summarize the discussion in that we should do two actions:

- [ ] Write up the Team Manual for code Management chapter (as the go-to place if your lab just produces a few analysis scripts)
- [ ] Finalise the https://github.com/the-turing-way/the-turing-way/issues/2419 (as the go-to place for larger projects)
- [ ] Point from one to the other (in both directions)

Does that sound like reasonable next steps?

**Comment by c-martinez** [comment]
Ok, so I think we could summarize the discussion in that we should do two actions:

- [ ] Write up the Team Manual for code Management chapter (as the go-to place if your lab just produces a few analysis scripts)
- [ ] Finalise the https://github.com/the-turing-way/the-turing-way/issues/2419 (as the go-to place for larger projects)
- [ ] Point from one to the other (in both directions)

Does that sound like reasonable next steps?

**Comment by flor14** [comment]
Thank you all for your comments in the thread.

- @hykelvinlee42: A very good resource. I was familiar with the Software Management Plan, but I didn't know there had been a focus on the vocabulary. A very valuable resource to cite and have as a reference. Thank you very much for sharing it!​

- @c-martinez: Good to know that you're working on this! Thank you for sharing all the links. I agree with addressing vocabulary questions together. I understand that the chapter you propose is broader. I believe the manual can be a useful tool in implementing a plan like the one you propose.

- @RichardJActon:  Your questions are appropriate to include as prompts in a manual. I believe the manual can serve both purposes: 
(1) as a straightforward way to introduce software management to researchers who are not yet familiar with the topic, and (2) as a means to help individuals transition from planning to action when implementing a software management plan. I agree that, as a subchapter of the lab manual, it can fulfill both functions, particularly (1).

- @srtee: I agree that the way the topic is introduced is key to engaging new audiences. The team software manual could include a section stating: "If you want to learn more, you can read about how to create software management plans." It would also be helpful to add a link to the page being developed by the RSE freelancers group, in case someone wants to consult with experts on how to develop something more advanced.

- @envismo @ml-evs: ​I hope you are doing well. I'm including you because a Turing Way Chapter might be a good place to add the page with the RSE freelance contacts. 

Feel free to leave the conversation if you can't participate, but feedback is appreciated 🙌🏼
